ADBDEV-2833: inject tableoid to orca targetlist for Update/Delete operation

HustonMmmavr opened the pull request:
## Patch description is inside the last commit.

## Problem description

Now, `Update/Delete` operation for partitioned tables and child ones [performs](https://github.com/greenplum-db/gpdb/blob/62b58c7634050be5c947e7151592725bd5fc82be/src/backend/executor/nodeModifyTable.c#L664-L681) hidden partition pruning that might lead to corruption when target relation from which tuple is extracted doesn't correspond to relation defined after partition pruning inside Update/Delete operation. That is, we perform removing of row using tupleid extracted from some lower node after scanning of right target relation but Update/Delete node doesn't have any info about that relation and tries to define it based on incoming partitioning key value via regular partition pruning routine. If target row to delete was located in wrong partition (e.g. after `ALTER TABLE EXCHANGE PARTITION WITHOUT VALIDATION` operation) then deletion will be issued using tupleid from that wrong partition on another right partition. This scenario in worth case ends up to deletion of some wrong row, i.e., data corruption.

For example:
```sql
-- Create partitioned table with one range and default partitions. Range partition is able to store two different values in partitioning key attribute. Now it stores one value.
create table test(i int, j int) partition by range(j) (start (1) end(3) every(2), default partition extra);
\d+ test
                         Table "public.test"
 Column |  Type   | Modifiers | Storage | Stats target | Description 
--------+---------+-----------+---------+--------------+-------------
 i      | integer |           | plain   |              | 
 j      | integer |           | plain   |              | 
Child tables: test_1_prt_2,
              test_1_prt_extra
Distributed by: (i)
Partition by: (j)
insert into test values (0, 1);

-- Create table for exchange with default partition and insert another possible value corresponding to range partition
create table test_extra_exchanged(like test);
insert into test_extra_exchanged values (0, 2);

-- Perform exchange without validation
set gp_enable_exchange_default_partition to on;
alter table test exchange default partition with table test_extra_exchanged without validation;

-- Check tuples in partitioned table. Two tuples have the same tupleid and value in partitioning key corresponding to `test_1_prt_2` partition
select tableoid::regclass, ctid, j from test;
     tableoid     | ctid  | j 
------------------+-------+---
 test_1_prt_extra | (0,1) | 2
 test_1_prt_2     | (0,1) | 1
(2 rows)

-- Create test table that will be used in delete query inside `NOT EXISTS` predicate
create table test_in_predicate(i int, j int);
insert into test_in_predicate values (0, 1);

-- We have to delete the row (0, 2) from default partition
explain (costs off, verbose)
delete from test where not exists(
  select 1 from test_in_predicate where test.i = test_in_predicate.i and test.j = test_in_predicate.j
);
                                             QUERY PLAN                                             
----------------------------------------------------------------------------------------------------
 Delete
   Output: test.i, test.j, "outer".ColRef_0026, test.ctid
   ->  Result
         Output: test.i, test.j, test.ctid, test.gp_segment_id, 0
         ->  Partition Selector for test
               Output: test.i, test.j, test.ctid, test.gp_segment_id
               ->  Hash Anti Join
                     Output: test.i, test.j, test.ctid, test.gp_segment_id
                     Hash Cond: ((test.i = test_in_predicate.i) AND (test.j = test_in_predicate.j))
                     ->  Sequence
                           Output: test.i, test.j, test.ctid, test.gp_segment_id
                           ->  Partition Selector for test (dynamic scan id: 1)
                                 Partitions selected: 2 (out of 2)
                           ->  Dynamic Seq Scan on public.test (dynamic scan id: 1)
                                 Output: test.i, test.j, test.ctid, test.gp_segment_id
                     ->  Hash
                           Output: "outer".?column?, test_in_predicate.i, test_in_predicate.j
                           ->  Result
                                 Output: 1, test_in_predicate.i, test_in_predicate.j
                                 ->  Seq Scan on public.test_in_predicate
                                       Output: test_in_predicate.i, test_in_predicate.j
 Optimizer: Pivotal Optimizer (GPORCA)
(22 rows)

delete from test where not exists(select 1 from test_in_predicate where test.i = test_in_predicate.i and test.j = test_in_predicate.j);
DELETE 1

-- But in fact the row from range partition was removed. This is corruption!
select tableoid::regclass, ctid, j from test;
     tableoid     | ctid  | j 
------------------+-------+---
 test_1_prt_extra | (0,1) | 2
(1 row)
```

The solution is based on patch from postgres which fixes the issue described above. When `Update/Delete` operation performs on the root table - `tableoid` value in extra colum is passed from the lower scanning node along with column containing tupleid value, this allows to detect partition correctly (like it's [done at postgres](https://github.com/postgres/postgres/blob/82699edbfe75534b3b6f6f4321339a432b7b8ff2/src/backend/executor/nodeModifyTable.c#L3533-L3572)).
In case of `Update/Delete` opeartion performed on the partition(section) there is no need in partition pruning, because the target relation already specified in metadata of execution state.

Also, ORCA part has changes too - partially returned work with `table Oid`, which was removed by the [PR 14578](https://github.com/greenplum-db/gpdb/pull/14578)

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

HustonMmmavr committed:
PG part: add tableoid to targetlist at rewriteHandler

tableoid added to targetlist only if table is  partitioned

HustonMmmavr committed:
ORCA: use tableoid at ORCA for partitioned tables

Pass tableoid through ORCA planner for delete/update operations
on the partitioned tables

- restored work with table oid inside planner nodes (like DML)
- fix the unit tests
- add hash of tableoid add serialization

HustonMmmavr committed:
PG part: add work with table oid for delete/update operations

Delete operation (in case of partitioned tables) now detects the
partition (table), where deletion should be performed by the `tableoid`
from the slot.tts_tuple, instead of calling `slot_get_partition` (this
function could detect the wrong partition (table) in case of default
partition exchaning). In case of updeate operation (which is a combination
of delete and insert operations) insert still uses `slot_get_partition`.

HustonMmmavr committed:
test: add sql and mdp tests

HustonMmmavr commented on src/backend/executor/nodeModifyTable.c:
```diff
@@ -658,26 +659,9 @@ ExecDelete(ItemPointer tupleid,
-#endif
-
-		Assert(parent != part);
+		resultRelInfo = targetid_get_partition(tableoid, estate, true);
```
- Maybe storing `resultRelation` to `estate` `estate->es_result_relation_info = resultRelInfo;`  should stay, like it's done fr `ExecInsert`. `estate->es_result_relation_info` must be rewrited to the `resultRelation` at `ExecInsert`, because it internally calls [`ExecInsertIndexTuples`](https://github.com/arenadata/gpdb/blob/adb-6.x-dev/src/backend/executor/nodeModifyTable.c#L574) which works only with `estate` and doesn't know anything about `resultRelation` declared at `ExecInsert`. In case of partitioned tables tests were failed - estate with incorrect `resultRelInfo` were passed) 
- `targetid_get_partition` - function called from `slot_get_partition`

HustonMmmavr replied:
refactored work with resultRelInfo from estate

maksm90 commented on src/backend/executor/nodeModifyTable.c:
```diff
@@ -658,26 +659,9 @@ ExecDelete(ItemPointer tupleid,
-	 * get information on the (current) result relation
-	 */
-	if (estate->es_result_partitions && planGen == PLANGEN_OPTIMIZER)
+	if (OidIsValid(tableoid) && planGen == PLANGEN_OPTIMIZER)
```
You might use ternary operator to reduce construction of `resultRelInfo` assignment

maksm90 commented on src/test/regress/sql/partition_pruning.sql:
```diff
@@ -973,4 +973,103 @@ where  trg.key1 = src.key1
+-- Tests of partition pruning for case when default partition was exchanged without validation
+-- and contains rows, that satisfy constraints of other partition
+
+-- simple delete test case
```
... delete test case from default partition. `Delete` operation has not to perform partition pruning

maksm90 commented on src/test/regress/sql/partition_pruning.sql:
```diff
@@ -973,4 +973,103 @@ where  trg.key1 = src.key1
+set gp_enable_exchange_default_partition to on;
+alter table test exchange default partition with table test_extra_exchanged without validation;
+
+select tableoid::regclass, ctid, i, j from test;
```
For current sql text and further please add comment lines about intention of sql blocks

HustonMmmavr replied:
done

maksm90 commented on src/backend/gporca/libgpopt/src/xforms/CXformUtils.cpp:
```diff
@@ -1340,7 +1338,7 @@ CXformUtils::PexprLogicalDMLOverProject(CMemoryPool *mp,
 	CExpression *pexprProject = NULL;
 	CColRef *pcrAction = NULL;
-	CColRef *pcrOid = NULL;
+	CColRef *pcrOid = pcrTableOid;
```
In this function we have to fully remove the logic related with definition of partition oid column as we define this column from outside

HustonMmmavr replied:
Removing logic of generation pcrOid for Insert operation (and using the pcrOid from the function argument (in case of insert it's always `NULL`)) leads to `SIGSEGV` at next test cases of `CDMLTest`.

- src/backend/gporca/data/dxl/minidump/Insert-Parquet-Partitioned.mdp
- src/backend/gporca/data/dxl/minidump/Insert-AO-Partitioned.mdp

pcrOid is `NULL` for insert operation and the problem occurs at this [code fragment (NULL pointer dereference)](https://github.com/arenadata/gpdb/blob/677a3cbda8b571709a35c5341d0024f9f38a63a6/src/backend/gporca/libgpopt/src/operators/CPhysicalDML.cpp#L531-L537).

- For the first test case `pcrOid` (previously) was generated , because it's a `parquet` table ([`FGeneratePartOid` returns true](https://github.com/arenadata/gpdb/blob/677a3cbda8b571709a35c5341d0024f9f38a63a6/src/backend/gporca/libgpopt/src/base/CUtils.cpp#L4507-L4525))
- For the second test case [Hint option is used](https://github.com/arenadata/gpdb/blob/677a3cbda8b571709a35c5341d0024f9f38a63a6/src/backend/gporca/data/dxl/minidump/Insert-AO-Partitioned.mdp#L14) and `FGeneratePartOid` returns `true`, while at adb-6.x [COptTask sets this hint to int_max](https://github.com/arenadata/gpdb/blob/677a3cbda8b571709a35c5341d0024f9f38a63a6/src/backend/gpopt/utils/COptTasks.cpp#L397).

So logic was changed - pcrOid is generated only for insert operation

maksm90 commented on src/backend/gporca/libgpopt/src/xforms/CXformUpdate2DML.cpp:
```diff
@@ -144,13 +144,20 @@ CXformUpdate2DML::Transform(CXformContext *pxfctxt, CXformResult *pxfres,
 	CExpression *pexprProject = NULL;
 	CColRef *pcrTableOid = NULL;
+	// seems this if statement should be changed, because at adb-6.x-dev plans - there are
+	// there are no info about tableoid, also slot at debug doesn't have this info
```
Hmm, this point needs some clarification

HustonMmmavr replied:
The logic of generation `oid` request seems strange, but if delete an old logic for partitioned tables (for update operation) and using the `pcrOid` from `LogicalUpdate` will broke some test cases from `CDMLTest` (when update operation doesn't have `pcrOid` info for exapmle replaying of plan from dxl without `OidCol`):
  - src/backend/gporca/data/dxl/minidump/UpdateDistKeyMismatchedDistribution.mdp
  - src/backend/gporca/data/dxl/minidump/UpdateDroppedCols.mdp
  - src/backend/gporca/data/dxl/minidump/UpdateNoDistKeyMismatchedDistribution.mdp
  - src/backend/gporca/data/dxl/minidump/UpdatePartTable.mdp

Here is an assertion (below in code) https://github.com/arenadata/gpdb/blob/762fcb30a9e4a1f0d9af1ff4d9f06fa88e5dc35c/src/backend/gporca/libgpopt/src/xforms/CXformUpdate2DML.cpp#L171
These test doesn't have `OidCol` at `dxlLogicalUpdate` node, so assertion fails.
These tests may be [fixed by adding OidCol](https://github.com/arenadata/gpdb/commit/0c670de15ec2c87718b447e785b2a932cff86b9f) to the `dxlLogicalUpdate` node, but seems here may be the next problem: replay of plan generated on previous versions of gpdb (where is no `OidCol` at `dxlLogicalUpdate` node) will fail.
There is two approaches to solve problem:
- remove the `assert` with fixing tests
-  still generate `pcrOid` if `Update` node doesn't have prcOid (this approach seems more suitable, because removing `assert` is not nice)

maksm90 commented:
Some comments to PR description:

> ... but Update/Delete node doesn't have any info about that target relation and tries to define it via internal partition pruning

.. but Update/Delete node doesn't have any info about that relation and tries to define it based on incoming partitioning key value via regular partition pruning routine.

> explain (costs off, verbose)
> delete from test where not exists(
>   select 1 from test_in_predicate where test.i = test_in_predicate.i and test.j = test_in_predicate.j
> );
>                                                       QUERY PLAN                                                       
> 
> 
> -----------------------------------------------------------------------------------------------------------------------
>  Delete  (cost=0.00..862.09 rows=1 width=1)
>    ->  Result  (cost=0.00..862.00 rows=1 width=22)
>    ...

show plan corresponding to requested parameters of `explain` - without costs, with explicit output of targetlist

> The solution is [based on patch from postgres](https://github.com/postgres/postgres/commit/86dc90056dfdbd9d1b891718d2e5614e3e432f35) which fixes the issue described above.

I think we haven't to refer to commit where among other things passing of tableoid value is added as the change is massive enough. Moreover, below you point the link to code fragment with this logic.

> ... already specified in query.

... already specified in metadata of execution state.

HustonMmmavr commented on src/backend/gporca/libgpopt/src/operators/CLogicalDelete.cpp:
```diff
@@ -122,6 +128,8 @@ CLogicalDelete::HashValue() const
 	ulHash = gpos::CombineHashes(ulHash, gpos::HashPtr<CColRef>(m_pcrCtid));
 	ulHash =
 		gpos::CombineHashes(ulHash, gpos::HashPtr<CColRef>(m_pcrSegmentId));
+	// does the calculation of tableoid is needed here ??? like at update
```
added calculation of hash value to LogicalDelete + LogicalUpdate, like it's done for LogicalDML

HustonMmmavr commented on src/backend/gporca/libgpopt/src/xforms/CXformUpdate2DML.cpp:
```diff
@@ -144,13 +144,14 @@ CXformUpdate2DML::Transform(CXformContext *pxfctxt, CXformResult *pxfres,
+	// there are no info about tableoid, also slot at debug doesn't have this info
 	if (ptabdesc->IsPartitioned())
 	{
+		pcrTableOid = popUpdate->PcrTableOid();
```
1  Deletion of [pcrOid generation at CXformUpdate2DML.cpp](https://github.com/arenadata/gpdb/pull/409/commits/05d6898c381640b9e6b13a9f99a9a09b89585458#diff-a44f4fd7c55d61e6605ae66bffda453c453e022468e3fbc443038adf1d3844a1) for partitioned tables (for update operation) and using the pcrOid from LogicalUpdate will broke some test cases from `CDMLTest`:
  - src/backend/gporca/data/dxl/minidump/UpdateDistKeyMismatchedDistribution.mdp
  - src/backend/gporca/data/dxl/minidump/UpdateDroppedCols.mdp
  - src/backend/gporca/data/dxl/minidump/UpdateNoDistKeyMismatchedDistribution.mdp
  - src/backend/gporca/data/dxl/minidump/UpdatePartTable.mdp

due to next [assertion]((https://github.com/arenadata/gpdb/blob/762fcb30a9e4a1f0d9af1ff4d9f06fa88e5dc35c/src/backend/gporca/libgpopt/src/xforms/CXformUpdate2DML.cpp#L171)) (test cases doesn't have OidCol at dxlLogicalUpdate node). These tests may be [fixed by adding OidCol](https://github.com/arenadata/gpdb/pull/409/commits/0c670de15ec2c87718b447e785b2a932cff86b9f) to the dxlLogicalUpdate node, but seems it will lead to problem with replay of plan generated on previous versions of gpdb (there is no OidCol so assert will fail). There is two approaches - remove the Assertion with fixing tests or generate pcrOid if Update node doesn't have prcOid.

HustonMmmavr commented on src/backend/gporca/libgpopt/src/xforms/CXformUtils.cpp:
```diff
@@ -1340,19 +1338,14 @@ CXformUtils::PexprLogicalDMLOverProject(CMemoryPool *mp,
 		// on top of that to add the action column
 		CExpression *pexprSelector = PexprLogicalPartitionSelector(
 			mp, ptabdesc, colref_array, pexprChild);
-		if (CUtils::FGeneratePartOid(ptabdesc->MDId()))
```
Deletion of generation pcrOid for Insert operation (and using the pcrOid from the function argument (in case of insert it's always NULL)) leads to SIGSEGV at next test cases of `CDMLTest`.

- src/backend/gporca/data/dxl/minidump/Insert-Parquet-Partitioned.mdp
- src/backend/gporca/data/dxl/minidump/Insert-AO-Partitioned.mdp

For now pcrOid is `NULL` and the problem occurrs at this [code](https://github.com/arenadata/gpdb/blob/677a3cbda8b571709a35c5341d0024f9f38a63a6/src/backend/gporca/libgpopt/src/operators/CPhysicalDML.cpp#L531-L537).

Previously for the first test case Oid was generated due to FGeneratePartOid check passed (here is a [function](https://github.com/arenadata/gpdb/blob/677a3cbda8b571709a35c5341d0024f9f38a63a6/src/backend/gporca/libgpopt/src/base/CUtils.cpp#L4410-L4429))
- first test case is called on parquet table.
-  second test case [uses special hint](https://github.com/arenadata/gpdb/blob/677a3cbda8b571709a35c5341d0024f9f38a63a6/data/dxl/minidump/Insert-AO-Partitioned.mdp#L14) while adb-6.x [COptTask sets this hint to int_max](https://github.com/arenadata/gpdb/blob/677a3cbda8b571709a35c5341d0024f9f38a63a6/src/backend/gpopt/utils/COptTasks.cpp#L397).

Seems this if should be returned

Commit:
Fix code style to pass linter test

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/59024

BenderArenadata commented:
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/810525

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/810526

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/810527

BenderArenadata commented:
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/810519

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/810518

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/810520

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/810521

Commit:
Reset environment to expected after last existed test

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/59074

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/815005

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/815002

BenderArenadata commented:
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/815010

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/815011

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/815012

bimboterminator1 commented on src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp:
```diff
@@ -1121,51 +1121,22 @@ CTranslatorQueryToDXL::ExtractStorageOptionStr(DefElem *def_elem)
-//		Obtains the id of the tuple oid column for the target table of a DML
-//		update
+//		returns the corresponding ColId for the given system
+//		attribute numbber
```
typo

InnerLife0 commented on src/backend/gporca/libgpopt/include/gpopt/operators/CLogicalUpdate.h:
```diff
@@ -46,9 +46,12 @@ class CLogicalUpdate : public CLogical
 	CColRef *m_pcrSegmentId;
 
-	// tuple oid column
+	// table oid column (has value NULL in case of on non partitioned tables)
```
I think this was changed by mistake and should be moved below.

bimboterminator1 replied:
Yep

InnerLife0 commented on src/backend/rewrite/rewriteHandler.c:
```diff
@@ -1411,6 +1413,19 @@ rewriteTargetListUD(Query *parsetree, RangeTblEntry *target_rte,
 						   type_coll,
 						   0);
+
+		if (rel_is_partitioned(reloid))
```
Can it be changed to avoid unnecessary job for postgres planner?

bimboterminator1 replied:
The problem is that service columns `ctid` and `gp_segment_id` are added to targetList at rewriting stage, and we need somehow to do the same only when relation is partitioned. And yep, this code leads to `tableoid` be added for legacy planner as well.  I think that is an unneseccary move, because there is no need for that column for postgres's plans. 
```
explain (costs off, verbose)  delete from test where j = 3;
LOG:  statement: explain (costs off, verbose)  delete from test where j = 3;
                                            QUERY PLAN                                            
--------------------------------------------------------------------------------------------------
 Delete on public.test_1_prt_extra
   ->  Seq Scan on public.test_1_prt_extra
         Output: test_1_prt_extra.ctid, test_1_prt_extra.gp_segment_id, test_1_prt_extra.tableoid
         Filter: (test_1_prt_extra.j = 3)
   ->  Seq Scan on public.test_1_prt_3
         Output: test_1_prt_3.ctid, test_1_prt_3.gp_segment_id, test_1_prt_3.tableoid
         Filter: (test_1_prt_3.j = 3)
 Optimizer: Postgres query optimizer
 Settings: optimizer=off
(9 rows)
```
Probably we could avoid that by extending the targetList only when ORCA is invoked.

bimboterminator1 replied:
At first glance, this code may be removed because (at Query to DXL translator stage)  `tableoid_colid` for CDXLLogicalDelete is taken (GetSystemColId) from `var_to_colid_mapping`, which is firstly filled [directly](https://github.com/arenadata/gpdb/blob/c6bbc225a01938db2ec08015a1a43e04e69371c5/src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp#L3284-L3285) from TableDescriptor, which definitely has tableoid.  Talking about other usegaes of targetlist entries, they also also fill the `output_attno_to_colid_mapping`.  In majority cases `output_attno_to_colid_mapping`  does not take into the account the`resjunk` attributes (for ex. [here](https://github.com/arenadata/gpdb/blob/c6bbc225a01938db2ec08015a1a43e04e69371c5/src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp#L1737-L1745), [here](https://github.com/arenadata/gpdb/blob/c6bbc225a01938db2ec08015a1a43e04e69371c5/src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp#L2584-L2588)). Considering the simple delete query `delete from test where j = 1`   all resjunk attributes ( ctid, gp_segment_id) are not extracted from it [here](https://github.com/arenadata/gpdb/blob/c6bbc225a01938db2ec08015a1a43e04e69371c5/src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp#L573-L574), and do not form any project list entries  [here](https://github.com/arenadata/gpdb/blob/c6bbc225a01938db2ec08015a1a43e04e69371c5/src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp#L4087-L4092) .
I'll probably remove this code. I guess for ORCA we won't need even ctid and gp_segment_id being in targetlist

bimboterminator1 replied:
```
explain (costs off, verbose)  delete from test where j = 3;
                                        QUERY PLAN                                        
------------------------------------------------------------------------------------------
 Delete
   Output: test.i, test.j, "outer".ColRef_0009, test.ctid, test.tableoid
   ->  Result
         Output: test.i, test.j, test.ctid, test.tableoid, test.gp_segment_id, 0
         ->  Sequence
               Output: test.i, test.j, test.ctid, test.tableoid, test.gp_segment_id
               ->  Partition Selector for test (dynamic scan id: 1)
                     Partitions selected: 1 (out of 3)
               ->  Dynamic Seq Scan on public.test (dynamic scan id: 1)
                     Output: test.i, test.j, test.ctid, test.tableoid, test.gp_segment_id
                     Filter: (test.j = 3)
 Optimizer: Pivotal Optimizer (GPORCA)
```
Let's consider this issue again.  `Tableoid_colid` for `CDXLLogicalDelete` is [taken](https://github.com/arenadata/gpdb/blob/2b3d7d92d43967b1df487b1a8d94ffdc379d661c/src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp#L1136-L1137) (GetSystemColId) from `var_to_colid_mapping`. If exploring the ways it's filled we get:
1) [RTE to DXLLogicalGET](https://github.com/arenadata/gpdb/blob/f6f95b6622a3fbdf441ec86164d5921b472694d7/src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp#L3284-L3286), the map is filled with references to Table Descriptor attributes, which correspond to base RTE_RELATION.
2) With entities corresponding to [Values Scan RTE,](https://github.com/arenadata/gpdb/blob/f6f95b6622a3fbdf441ec86164d5921b472694d7/src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp#L3583-L3585) not our case
3) And [various other types of RTE ](https://github.com/arenadata/gpdb/blob/f6f95b6622a3fbdf441ec86164d5921b472694d7/src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp#L3157-L3162), which are definetely not our case, because [m_query->resultRelation](https://github.com/arenadata/gpdb/blob/2b3d7d92d43967b1df487b1a8d94ffdc379d661c/src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp#L1136-L1137) may [point](https://github.com/arenadata/gpdb/blob/f6f95b6622a3fbdf441ec86164d5921b472694d7/src/backend/parser/analyze.c#L3034) only to RTE_RELATION, what means that `m_var_to_colid_mapping` may be filled with Colid corresponding to the RTE_RELATION (see the point 1.)

It means, that top-level target list of the Query, which is considered in this review comment, does not participate in var_to_colid_mapping constrction.

Second point we want to be sure about is whether the final plan's target lists are filled with tableoid. When performing optimization jobs for CGroupExpression, containing CPhysicalDML operator, we shedule optimization jobs for CGroupExpression's children. During that process, we [compute](https://github.com/arenadata/gpdb/blob/f6f95b6622a3fbdf441ec86164d5921b472694d7/src/backend/gporca/libgpopt/src/search/CJobGroupExpressionOptimization.cpp#L478) required properties for children based on required properies, that come on parent from above (it means that there were required properties expected from parent to provide). For the CGroupExpression with CPhysicalDML we shedule the jobs for its children and compute required properties for them based on properties of CPhysicalDML. The compute func [calls](https://github.com/arenadata/gpdb/blob/f6f95b6622a3fbdf441ec86164d5921b472694d7/src/backend/gporca/libgpopt/src/base/CReqdPropPlan.cpp#L174) `ComputeReqdCols`, which [calls](https://github.com/arenadata/gpdb/blob/f6f95b6622a3fbdf441ec86164d5921b472694d7/src/backend/gporca/libgpopt/src/base/CReqdPropPlan.cpp#L131-L132) `PcrsRequired` , on CPhysicalDML [operator](https://github.com/arenadata/gpdb/blob/f6f95b6622a3fbdf441ec86164d5921b472694d7/src/backend/gporca/libgpopt/src/operators/CPhysicalDML.cpp#L215-L234), which retrns column set, [containing](https://github.com/arenadata/gpdb/blob/2b3d7d92d43967b1df487b1a8d94ffdc379d661c/src/backend/gporca/libgpopt/src/operators/CPhysicalDML.cpp#L546-L549) our table oid. Starting from that point, optimization process propagates the requirement of this column to child groups. 
How does this stuff get to the target list? 
Consider translation from Expr to DXL
1) Root CDXLPhysicalDML [ gets tableoid as it shown in our patch](https://github.com/arenadata/gpdb/blob/2b3d7d92d43967b1df487b1a8d94ffdc379d661c/src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp#L5796-L5798), project list of future DXL node is [combined with familiar required columns and output columns](https://github.com/arenadata/gpdb/blob/2b3d7d92d43967b1df487b1a8d94ffdc379d661c/src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp#L5801-L5802) (non-system attributes)
2)  Recursion down to the bottom DynamicTableScan with PartitionSelector. The DXL node projection list is [created by the similar combination of required and table output columns](https://github.com/arenadata/gpdb/blob/2b3d7d92d43967b1df487b1a8d94ffdc379d661c/src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp#L1118-L1120).
3) Recurse back from bottom, and see, that each projection list is created the same way, except Partiton Selector, it does not have projection list, we can see that in the vebose explain output.
At this point we have projection lists filled:
<details>
  <summary>DXL PLAN</summary>
  
```
<dxl:Plan Id="0" SpaceSize="2">
      <dxl:DMLDelete Columns="0,1" ActionCol="9" CtidCol="2" OidCol="7" SegmentIdCol="8">
        <dxl:Properties>
          <dxl:Cost StartupCost="0" TotalCost="431.101683" Rows="1.000000" Width="1"/>
        </dxl:Properties>
        <dxl:DirectDispatchInfo/>
        <dxl:ProjList>
          <dxl:ProjElem ColId="0" Alias="i">
            <dxl:Ident ColId="0" ColName="i" TypeMdid="0.23.1.0"/>
          </dxl:ProjElem>
          <dxl:ProjElem ColId="1" Alias="j">
            <dxl:Ident ColId="1" ColName="j" TypeMdid="0.23.1.0"/>
          </dxl:ProjElem>
        </dxl:ProjList>
        <dxl:TableDescriptor Mdid="6.16420.1.0" TableName="test">
          <dxl:Columns>
            <dxl:Column ColId="10" Attno="1" ColName="i" TypeMdid="0.23.1.0" ColWidth="4"/>
            <dxl:Column ColId="11" Attno="2" ColName="j" TypeMdid="0.23.1.0" ColWidth="4"/>
            <dxl:Column ColId="12" Attno="-1" ColName="ctid" TypeMdid="0.27.1.0" ColWidth="6"/>
            <dxl:Column ColId="13" Attno="-3" ColName="xmin" TypeMdid="0.28.1.0" ColWidth="4"/>
            <dxl:Column ColId="14" Attno="-4" ColName="cmin" TypeMdid="0.29.1.0" ColWidth="4"/>
            <dxl:Column ColId="15" Attno="-5" ColName="xmax" TypeMdid="0.28.1.0" ColWidth="4"/>
            <dxl:Column ColId="16" Attno="-6" ColName="cmax" TypeMdid="0.29.1.0" ColWidth="4"/>
            <dxl:Column ColId="17" Attno="-7" ColName="tableoid" TypeMdid="0.26.1.0" ColWidth="4"/>
            <dxl:Column ColId="18" Attno="-8" ColName="gp_segment_id" TypeMdid="0.23.1.0" ColWidth="4"/>
          </dxl:Columns>
        </dxl:TableDescriptor>
        <dxl:Result>
          <dxl:Properties>
            <dxl:Cost StartupCost="0" TotalCost="431.000121" Rows="1.000000" Width="26"/>
          </dxl:Properties>
          <dxl:ProjList>
            <dxl:ProjElem ColId="0" Alias="i">
              <dxl:Ident ColId="0" ColName="i" TypeMdid="0.23.1.0"/>
            </dxl:ProjElem>
            <dxl:ProjElem ColId="1" Alias="j">
              <dxl:Ident ColId="1" ColName="j" TypeMdid="0.23.1.0"/>
            </dxl:ProjElem>
            <dxl:ProjElem ColId="2" Alias="ctid">
              <dxl:Ident ColId="2" ColName="ctid" TypeMdid="0.27.1.0"/>
            </dxl:ProjElem>
            <dxl:ProjElem ColId="7" Alias="tableoid">
              <dxl:Ident ColId="7" ColName="tableoid" TypeMdid="0.26.1.0"/>
            </dxl:ProjElem>
            <dxl:ProjElem ColId="8" Alias="gp_segment_id">
              <dxl:Ident ColId="8" ColName="gp_segment_id" TypeMdid="0.23.1.0"/>
            </dxl:ProjElem>
            <dxl:ProjElem ColId="9" Alias="ColRef_0009">
              <dxl:ConstValue TypeMdid="0.23.1.0" Value="0"/>
            </dxl:ProjElem>
          </dxl:ProjList>
          <dxl:Filter/>
          <dxl:OneTimeFilter/>
          <dxl:Sequence>
            <dxl:Properties>
              <dxl:Cost StartupCost="0" TotalCost="431.000095" Rows="1.000000" Width="22"/>
            </dxl:Properties>
            <dxl:ProjList>
              <dxl:ProjElem ColId="0" Alias="i">
                <dxl:Ident ColId="0" ColName="i" TypeMdid="0.23.1.0"/>
              </dxl:ProjElem>
              <dxl:ProjElem ColId="1" Alias="j">
                <dxl:Ident ColId="1" ColName="j" TypeMdid="0.23.1.0"/>
              </dxl:ProjElem>
              <dxl:ProjElem ColId="2" Alias="ctid">
                <dxl:Ident ColId="2" ColName="ctid" TypeMdid="0.27.1.0"/>
              </dxl:ProjElem>
              <dxl:ProjElem ColId="7" Alias="tableoid">
                <dxl:Ident ColId="7" ColName="tableoid" TypeMdid="0.26.1.0"/>
              </dxl:ProjElem>
              <dxl:ProjElem ColId="8" Alias="gp_segment_id">
                <dxl:Ident ColId="8" ColName="gp_segment_id" TypeMdid="0.23.1.0"/>
              </dxl:ProjElem>
            </dxl:ProjList>
            <dxl:PartitionSelector RelationMdid="6.16420.1.0" PartitionLevels="1" ScanId="1">
              <dxl:Properties>
                <dxl:Cost StartupCost="10" TotalCost="100" Rows="100" Width="4"/>
              </dxl:Properties>
              <dxl:ProjList/>
              <dxl:PartEqFilters>
                <dxl:PartEqFilterElems>
                  <dxl:ConstValue TypeMdid="0.23.1.0" Value="3"/>
                </dxl:PartEqFilterElems>
              </dxl:PartEqFilters>
              <dxl:PartFilters>
                <dxl:ConstValue TypeMdid="0.16.1.0" Value="true"/>
              </dxl:PartFilters>
              <dxl:ResidualFilter>
                <dxl:ConstValue TypeMdid="0.16.1.0" Value="true"/>
              </dxl:ResidualFilter>
              <dxl:PropagationExpression>
                <dxl:ConstValue TypeMdid="0.23.1.0" Value="1"/>
              </dxl:PropagationExpression>
              <dxl:PrintableFilter>
                <dxl:Comparison ComparisonOperator="=" OperatorMdid="0.96.1.0">
                  <dxl:Ident ColId="1" ColName="j" TypeMdid="0.23.1.0"/>
                  <dxl:ConstValue TypeMdid="0.23.1.0" Value="3"/>
                </dxl:Comparison>
              </dxl:PrintableFilter>
            </dxl:PartitionSelector>
            <dxl:DynamicTableScan PartIndexId="1">
              <dxl:Properties>
                <dxl:Cost StartupCost="0" TotalCost="431.000095" Rows="1.000000" Width="22"/>
              </dxl:Properties>
              <dxl:ProjList>
                <dxl:ProjElem ColId="0" Alias="i">
                  <dxl:Ident ColId="0" ColName="i" TypeMdid="0.23.1.0"/>
                </dxl:ProjElem>
                <dxl:ProjElem ColId="1" Alias="j">
                  <dxl:Ident ColId="1" ColName="j" TypeMdid="0.23.1.0"/>
                </dxl:ProjElem>
                <dxl:ProjElem ColId="2" Alias="ctid">
                  <dxl:Ident ColId="2" ColName="ctid" TypeMdid="0.27.1.0"/>
                </dxl:ProjElem>
                <dxl:ProjElem ColId="7" Alias="tableoid">
                  <dxl:Ident ColId="7" ColName="tableoid" TypeMdid="0.26.1.0"/>
                </dxl:ProjElem>
                <dxl:ProjElem ColId="8" Alias="gp_segment_id">
                  <dxl:Ident ColId="8" ColName="gp_segment_id" TypeMdid="0.23.1.0"/>
                </dxl:ProjElem>
              </dxl:ProjList>
              <dxl:Filter>
                <dxl:Comparison ComparisonOperator="=" OperatorMdid="0.96.1.0">
                  <dxl:Ident ColId="1" ColName="j" TypeMdid="0.23.1.0"/>
                  <dxl:ConstValue TypeMdid="0.23.1.0" Value="3"/>
                </dxl:Comparison>
              </dxl:Filter>
              <dxl:TableDescriptor Mdid="6.16420.1.0" TableName="test">
                <dxl:Columns>
                  <dxl:Column ColId="0" Attno="1" ColName="i" TypeMdid="0.23.1.0" ColWidth="4"/>
                  <dxl:Column ColId="1" Attno="2" ColName="j" TypeMdid="0.23.1.0" ColWidth="4"/>
                  <dxl:Column ColId="2" Attno="-1" ColName="ctid" TypeMdid="0.27.1.0" ColWidth="6"/>
                  <dxl:Column ColId="3" Attno="-3" ColName="xmin" TypeMdid="0.28.1.0" ColWidth="4"/>
                  <dxl:Column ColId="4" Attno="-4" ColName="cmin" TypeMdid="0.29.1.0" ColWidth="4"/>
                  <dxl:Column ColId="5" Attno="-5" ColName="xmax" TypeMdid="0.28.1.0" ColWidth="4"/>
                  <dxl:Column ColId="6" Attno="-6" ColName="cmax" TypeMdid="0.29.1.0" ColWidth="4"/>
                  <dxl:Column ColId="7" Attno="-7" ColName="tableoid" TypeMdid="0.26.1.0" ColWidth="4"/>
                  <dxl:Column ColId="8" Attno="-8" ColName="gp_segment_id" TypeMdid="0.23.1.0" ColWidth="4"/>
                </dxl:Columns>
              </dxl:TableDescriptor>
            </dxl:DynamicTableScan>
          </dxl:Sequence>
        </dxl:Result>
      </dxl:DMLDelete>
    </dxl:Plan>

```

</details>

After the projection lists are set, [we fill the planned stmt's target lists from them](https://github.com/arenadata/gpdb/blob/2b3d7d92d43967b1df487b1a8d94ffdc379d661c/src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp#L4754).

InnerLife0 commented on src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:
```diff
@@ -4191,6 +4191,21 @@ CTranslatorDXLToPlStmt::TranslateDXLDml(
+	else if (md_rel->IsPartitioned() &&
+			 (CMD_UPDATE == m_cmd_type || CMD_DELETE == m_cmd_type))
+	{
+		GPOS_RAISE(
```
Dunno if we should be so aggressive here. ORCA can plan a query even without tableoid. The error is more related to executor.

bimboterminator1 replied:
Totally agree. If we are extending optimizer's behaviour we should guarantee the presence of `tableoid`. And yep, it's better to move  this check to executor level.

bimboterminator1 replied:
On the other hand,  we will cut off an opportunity of falling back to the postgres planner. There can be a very specific case when this patch won't work (I couldn't find one).

InnerLife0 commented:
Is it a bug or feature? :)
```
postgres=# create table test(i int, j int) partition by range(j) (start (1) end(5) every(2), default partition extra);
CREATE TABLE
postgres=# 
postgres=# -- check that insert into partition table won't fallback in pg optimizer and won't request tableoid
postgres=# explain verbose insert into test values (0, 1);
                            QUERY PLAN                            
------------------------------------------------------------------
 Insert  (cost=0.00..0.02 rows=1 width=8)
   Output: i, j, ColRef_0003
   ->  Result  (cost=0.00..0.00 rows=1 width=12)
         Output: i, j, 1
         ->  Result  (cost=0.00..0.00 rows=1 width=8)
               Output: i, j
               ->  Result  (cost=0.00..0.00 rows=1 width=8)
                     Output: 0, 1
                     ->  Result  (cost=0.00..0.00 rows=1 width=1)
                           Output: true
 Optimizer: Pivotal Optimizer (GPORCA)
 Settings: optimizer=on
(12 rows)

postgres=# insert into test values (0, 1);
INSERT 0 1
postgres=# insert into test values (0, 3);
INSERT 0 1
postgres=# -- Create table for exchange with default partition and insert another possible
postgres=# -- value corresponding to range partition
postgres=# create table test_extra_exchanged(like test);
CREATE TABLE
postgres=# insert into test_extra_exchanged values (0, 1);
INSERT 0 1
postgres=# 
postgres=# -- Perform exchange without validation
postgres=# --set gp_enable_exchange_default_partition to on;
postgres=# alter table test exchange partition for ('3') with table test_extra_exchanged without validation;
ALTER TABLE
postgres=# 
postgres=# -- Check tuples in partitioned table. Two tuples have the same tupleid and value in partitioning
postgres=# -- key corresponding to `test_1_prt_2` partition
postgres=# select tableoid::regclass, ctid, i, j from test;
   tableoid   | ctid  | i | j 
--------------+-------+---+---
 test_1_prt_2 | (0,1) | 0 | 1
 test_1_prt_3 | (0,1) | 0 | 1
(2 rows)

postgres=# 
postgres=# -- We have to delete the row (0, 2) from default partition
postgres=# explain (verbose, costs off) delete from test_1_prt_3 where j = 1;
                                                                   QUERY PLAN                                                                    
-------------------------------------------------------------------------------------------------------------------------------------------------
 Delete
   Output: i, j, ColRef_0009, ctid
   ->  Result
         Output: i, j, ctid, gp_segment_id, 0
         ->  Result
               Output: i, j, ctid, gp_segment_id
               ->  Result
                     Output: i, j, ctid, gp_segment_id
                     ->  Result
                           Output: NULL::integer, NULL::integer, NULL::tid, NULL::xid, NULL::cid, NULL::xid, NULL::cid, NULL::oid, NULL::integer
                           One-Time Filter: false
 Optimizer: Pivotal Optimizer (GPORCA)
 Settings: optimizer=on
(13 rows)

postgres=# delete from test_1_prt_3 where j = 1;
DELETE 0
postgres=# 
postgres=# -- Check that deletion performed correctly
postgres=# select tableoid::regclass, ctid, i, j from test;
   tableoid   | ctid  | i | j 
--------------+-------+---+---
 test_1_prt_2 | (0,1) | 0 | 1
 test_1_prt_3 | (0,1) | 0 | 1
(2 rows)

postgres=# 
postgres=# drop table test, test_extra_exchanged;
DROP TABLE
```
How can we delete a single row from test_1_prt_3?

bimboterminator1 commented on src/backend/gporca/libgpopt/include/gpopt/operators/CLogicalDML.h:
```diff
@@ -58,6 +58,10 @@ class CLogicalDML : public CLogical
+	// table oid column (has value NULL for Update/Delete operations on non partitioned
+	// tables and for all Insert operations)
+	CColRef *m_pcrTableOid;
+
```
Why have  we ignored `CLogicalPartitionSelector` and `CPhysicalPartitionSelector`? Both operators had the `pcrTableOid` in their definition according to the upstream patch. For example
[src/backend/gporca/libgpopt/include/gpopt/operators/CLogicalPartitionSelector.h](https://github.com/greenplum-db/gpdb/pull/14578/files#diff-e55bb03e729228d46d366b7c790e5680a9013491e6ebd10e153371b9d1ee6ad8)

bimboterminator1 replied:
These operators are created only at property enforcement stages and they eventually get `tableoid` column through [CReqdPropPlan::ComputeReqdCols](https://github.com/greenplum-db/gpdb/blob/0ed18a56fd5744e5a6c7806cd380db3464524c6c/src/backend/gporca/libgpopt/src/base/CReqdPropPlan.cpp#L114-L125)

bimboterminator1 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-2833-v2

bimboterminator1 committed:
Remove the tableoid target entry from Query struct

We may avoid adding tableoid attribute from targetlist of input Query
node because targetlist does not participate in Query DXL tree creation.
tableoid_colid for CDXLLogicalDelete is taken (GetSystemColId) from
var_to_colid_mapping, which is firstly filled directly from TableDescriptor,
which is created at CDXLLogicalSelect stage and which definitely has tableoid

bimboterminator1 committed:
Fix comment typos

bimboterminator1 committed:
Refactor ExecDelete's part with throwing and error

The error is thrown in case when deletion is taken on root partition,
but the tubleoid is absent.

execDML is changed with adding an additional condition in the
if clause because on DELETE the partition selection is not performed.

bimboterminator1 committed:
Change test oputput for planner

bimboterminator1 commented on src/backend/executor/nodeModifyTable.c:
```diff
@@ -671,31 +672,9 @@ ExecDelete(ItemPointer tupleid,
-
-		/* Obtain part for current tuple. */
-		resultRelInfo = slot_get_partition(planSlot, estate);
-		estate->es_result_relation_info = resultRelInfo;
```
We should take it back anyway. For DELETE it **safe** to remove that line, **but** the we should support consistency with INSERT and UPDATE (for them its crucial to store the pointer to modified relation, at least because of ExecInsertIndexTuples, which uses it). Have done that in the v2 [draft](https://github.com/arenadata/gpdb/pull/677).

bimboterminator1 committed:
Remove error from ExecDelete and add es_result_partition assignment

bimboterminator1 committed:
Remove header from rewriteHandler.c

bimboterminator1 committed:
Fix tests

bimboterminator1 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-2833-v1

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1049241

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1049242

BenderArenadata commented:
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1049234

bimboterminator1 committed:
Add tableoid attribute to ORCA UPDATE/DELETE targetlist

ORCA's Split Update and Delete DML operations performed partition pruning for
each incoming tuple at delete stage. Executor always called slot_get_partition
to get the target partition to delete from. This behaviour did cause not only
the increased computational cost, but it could lead to data corruption when the
relation, from which tuple is extracted, did not correspond to relation defined
after partition pruning inside Update/Delete operations. For example, if target
row to delete was located in wrong partition (e.g. after ALTER TABLE EXCHANGE
PARTITION WITHOUT VALIDATION operation), then deletion will be issued using
ctid from that wrong partition on another right partition (which was chosen from
slot_get_partition). This scenario in worth case ends up to deletion of some
wrong row, i.e., data corruption.

This patch proposes solution of the issue described above. Instead of performing
partition selection at delete stage in ExecDelete, it would be more optimal to
add a tableoid attribite to the incoming slot. This attribute will contain the
oid of the leaf partition, which the tuple actually belongs to. In this case
there won't be any need to repeat the partiton pruning, because the tuple for
Update/Delete will already contain the oid of its relation. In order to achive
that the patch adds the tableoid atrribute to the ORCA's plan target lists for
Update/Delete plans. To add the new attribute to the final target list, this
patch adds a tableoid column reference both to DXL DML nodes (Query and Plan
DXL) and to each logical and physical DML operator. We must notice that tableoid
is added only for DML operations executed on root partitions (partitioned
tables). The corresponding Xforms and translators are also extended with the
tableoid handling. Because the new column have been added to the operator
classes, the parser classes (CParseHandler*) have been also extended in order
to support the proper print output for either debug print or minidumps.

On executor side the patch removes the partition pruning code from the
ExecDelete function. Is uses targetid_get_partition with tableoid attribute of
the slot, which is extracted in ExecDML function.

The tests have been added both to the regress and ORCA's minidump suites. Added
minidumps show the presence of tableoid attribute in DXL plan. Regress tests
check the correctnes of DML operations for modifying tuples belonging to wrong
partitions as it is described above.

This patch is also modifies some code and tests from commit fd5939d in order to
adjust it to the tableoid presence.

bimboterminator1 committed:
Add tableoid attribute to ORCA UPDATE/DELETE targetlist

ORCA's Split Update and Delete DML operations performed partition pruning for
each incoming tuple at delete stage. Executor always called slot_get_partition
to get the target partition to delete from. This behaviour did cause not only
the increased computational cost, but it could lead to data corruption when the
relation, from which tuple is extracted, did not correspond to relation defined
after partition pruning inside Update/Delete operations. For example, if target
row to delete was located in wrong partition (e.g. after ALTER TABLE EXCHANGE
PARTITION WITHOUT VALIDATION operation), then deletion will be issued using
ctid from that wrong partition on another right partition (which was chosen from
slot_get_partition). This scenario in worth case ends up to deletion of some
wrong row, i.e., data corruption.

This patch proposes solution of the issue described above. Instead of performing
partition selection at delete stage in ExecDelete, it would be more optimal to
add a tableoid attribite to the incoming slot. This attribute will contain the
oid of the leaf partition, which the tuple actually belongs to. In this case
there won't be any need to repeat the partiton pruning, because the tuple for
Update/Delete will already contain the oid of its relation. In order to achive
that the patch adds the tableoid atrribute to the ORCA's plan target lists for
Update/Delete plans. To add the new attribute to the final target list, this
patch adds a tableoid column reference both to DXL DML nodes (Query and Plan
DXL) and to each logical and physical DML operator. We must notice that tableoid
is added only for DML operations executed on root partitions (partitioned
tables). The corresponding Xforms and translators are also extended with the
tableoid handling. Because the new column have been added to the operator
classes, the parser classes (CParseHandler*) have been also extended in order
to support the proper print output for either debug print or minidumps.

On executor side the patch removes the partition pruning code from the
ExecDelete function. Is uses targetid_get_partition with tableoid attribute of
the slot, which is extracted in ExecDML function.

The tests have been added both to the regress and ORCA's minidump suites. Added
minidumps show the presence of tableoid attribute in DXL plan. Regress tests
check the correctnes of DML operations for modifying tuples belonging to wrong
partitions as it is described above.

This patch also modifies some code and tests from commit fd5939d in order to
adjust it to the tableoid presence.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/64282

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1062458

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1062459

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1062456

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1062450

bimboterminator1 committed:
Remove tableoid from ExecDelete.Fix compatibility tests

bimboterminator1 committed:
Add tableoid attribute to ORCA UPDATE/DELETE targetlist

ORCA's Split Update and Delete DML operations performed partition pruning for
each incoming tuple at delete stage. Executor always called slot_get_partition
to get the target partition to delete from. This behaviour did cause not only
the increased computational cost, but it could lead to data corruption when the
relation, from which tuple is extracted, did not correspond to relation defined
after partition pruning inside Update/Delete operations. For example, if target
row to delete was located in wrong partition (e.g. after ALTER TABLE EXCHANGE
PARTITION WITHOUT VALIDATION operation), then deletion will be issued using
ctid from that wrong partition on another right partition (which was chosen from
slot_get_partition). This scenario in worth case ends up to deletion of some
wrong row, i.e., data corruption.

This patch proposes solution of the issue described above. Instead of performing
partition selection at delete stage in ExecDelete, it would be more optimal to
add a tableoid attribite to the incoming slot. This attribute will contain the
oid of the leaf partition, which the tuple actually belongs to. In this case
there won't be any need to repeat the partiton pruning, because the tuple for
Update/Delete will already contain the oid of its relation. In order to achive
that the patch adds the tableoid atrribute to the ORCA's plan target lists for
Update/Delete plans. To add the new attribute to the final target list, this
patch adds a tableoid column reference both to DXL DML nodes (Query and Plan
DXL) and to each logical and physical DML operator. We must notice that tableoid
is added only for DML operations executed on root partitions (partitioned
tables). The corresponding Xforms and translators are also extended with the
tableoid handling. Because the new column have been added to the operator
classes, the parser classes (CParseHandler*) have been also extended in order
to support the proper print output for either debug print or minidumps.

On executor side the patch removes the partition pruning code from the
ExecDelete function. Instead, we fill the estate->es_result_info with the result
of targetid_get_partition function, which creates the structure corresponding to
a leaf partition using tableoid attribute of the slot.

The tests have been added both to the regress and ORCA's minidump suites. Added
minidumps show the presence of tableoid attribute in DXL plan. Regress tests
check the correctnes of DML operations for modifying tuples belonging to wrong
partitions as it is described above.

This patch also modifies some code and tests from commit fd5939d in order to
adjust it to the tableoid presence.

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1064869

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1064870

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1064861

Commit:
Merge branch 'adb-6.x-dev' into ADBDEV-2833-v1

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/64871

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1082765

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1082766
